pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -339,7 +340,13 @@ func (m *ManagerImpl) Allocate(pod *v1.Pod, container *v1.Container) error {
 			if err := m.allocateContainerResources(pod, container, m.devicesToReuse[string(pod.UID)]); err != nil {
 				return err
 			}
-			m.podDevices.addContainerAllocatedResources(string(pod.UID), container.Name, m.devicesToReuse[string(pod.UID)])
+			if !types.IsRestartableInitContainer(&initContainer) {
+				m.podDevices.addContainerAllocatedResources(string(pod.UID), container.Name, m.devicesToReuse[string(pod.UID)])
+			} else {
+				// If the init container is restartable, we need to keep the
+				// devices allocated
```

Comment:
Can we add add more details here and explain the logic as it seemed counter intuitive to me at first (because in the comment we say said that we are keeping the devices but "appear" to be removing them).

Let's explain that for restartable init containers, we want to ensure that the devices once allocated to the container are not added back to `podDevices` unlike non-restartable init containers where we reclaim the resources from the containers and add them back to `podDevices`.

Reply:
Thanks for the review. I added some comments to make it clear.